lib/Sema/ConstraintSystem.h

```diff
@@ -3288,28 +3288,72 @@ class ConstraintSystem {
     /// The kind of bindings permitted.
     AllowedBindingKind Kind;
 
-    /// The kind of the constraint this binding came from.
-    ConstraintKind BindingSource;
-
-    /// The defaulted protocol associated with this binding.
-    ProtocolDecl *DefaultedProtocol;
-
-    /// If this is a binding that comes from a \c Defaultable constraint,
-    /// the locator of that constraint.
-    ConstraintLocator *DefaultableBinding = nullptr;
+  protected:
+    /// The source of the type information.
+    ///
+    /// Determines whether this binding represents a "hole" in
+    /// constraint system. Such bindings have no originating constraint
+    /// because they are synthetic, they have a locator instead.
+    PointerUnion<Constraint *, ConstraintLocator *> BindingSource;
 
     PotentialBinding(Type type, AllowedBindingKind kind,
-                     ConstraintKind bindingSource,
-                     ProtocolDecl *defaultedProtocol = nullptr,
-                     ConstraintLocator *defaultableBinding = nullptr)
+                     PointerUnion<Constraint *, ConstraintLocator *> source)
+        : BindingType(type->getWithoutParens()), Kind(kind),
+          BindingSource(source) {}
+
+  public:
+    PotentialBinding(Type type, AllowedBindingKind kind, Constraint *source)
         : BindingType(type->getWithoutParens()), Kind(kind),
-          BindingSource(bindingSource), DefaultedProtocol(defaultedProtocol),
-          DefaultableBinding(defaultableBinding) {}
+          BindingSource(source) {}
+
+    bool isDefaultableBinding() const {
+      if (auto *constraint = BindingSource.dyn_cast<Constraint *>())
+        return constraint->getKind() == ConstraintKind::Defaultable;
+      // If binding source is not constraint - it's a hole, which is
```

Comment:
So a "hole" is a type variable that couldn't have a fixed inferred by a constraint in the system? I'm just asking because I don't fully understand what "hole" is :)

Reply:
Yeah, a hole is a type variable which we couldn't infer a type for e.g. `_ = .foo` here there is no way to infer implicit base type (represented by `$T_base.Type`) for example, or something like this:

```
func foo<T>() -> T {}
_ = foo()
```

where there is no binding for generic parameter `T`.

Reply:
Humm got it. Thanks :)

Reply:
No problem!

Reply:
Is it related to typed holes in Haskell? Just curious to learn a bit more about how holes work

Reply:
Yes, that's pretty one 1-to-1 comparison.

Reply:
The only difference is that here this mechanism is only used for diagnostics at the moment but it could be extended to cover e.g. code completion as well.

Reply:
They're the same in the sense that they're placeholders for something that is yet to be specified (by the programmer). The difference between typed holes is Haskell and "holes" here is that in Haskell, a hole is a placeholder for another expression. Here, it's a placeholder for a type (that we couldn't infer)

Reply:
I think in Swift case as well it would be appropriate to consider holes as placeholders for missing expressions in cases like contextual base or missing argument(s) etc.

Reply:
What I mean is that in Haskell, a typed hole is something the programmer writes explicitly in source code in an expression context

Reply:
Ah yes, that’s definitely the difference here (for now) :)